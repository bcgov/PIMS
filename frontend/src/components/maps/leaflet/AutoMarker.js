import { Marker } from 'react-leaflet';
import React from 'react';

// Create your own class, extending from the Marker class.
const AutoMarker = props => {

  const openPopup = marker => {
    if (marker && props.autoPopup) {
        marker.leafletElement.openPopup();
    }
  };

  return (
    <Marker ref={el => openPopup(el)} {...props}/>
  );
};
export default AutoMarker;