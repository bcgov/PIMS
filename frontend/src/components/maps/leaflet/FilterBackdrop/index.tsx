import * as React from 'react';
import { Spinner } from 'react-bootstrap';
import styled from 'styled-components';

export interface FilterBackdropProps {
  show?: boolean;
}

const Backdrop = styled.div`
  width: 100%;
  height: 100%;
  position: fixed;
  z-index: 999;
  left: 0;
  background-color: rgba(0, 0, 0, 0.4);
  display: flex;
  align-items: center;
  align-content: center;
  justify-items: center;
  justify-content: center;
`;

const FilterBackdrop: React.FC<FilterBackdropProps> = ({ show }) => {
  return show ? (
    <Backdrop>
      <Spinner animation="border" variant="warning" />
    </Backdrop>
  ) : null;
};

export default FilterBackdrop;
