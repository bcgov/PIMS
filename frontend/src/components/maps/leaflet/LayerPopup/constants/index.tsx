import * as React from 'react';

export const MUNICIPALITY_LAYER_URL =
  'https://openmaps.gov.bc.ca/geo/pub/WHSE_LEGAL_ADMIN_BOUNDARIES.ABMS_MUNICIPALITIES_SP/wfs?SERVICE=WFS&REQUEST=GetFeature&VERSION=1.3.0&outputFormat=application/json&typeNames=pub:WHSE_LEGAL_ADMIN_BOUNDARIES.ABMS_MUNICIPALITIES_SP';
export const PARCELS_LAYER_URL =
  'https://openmaps.gov.bc.ca/geo/pub/WHSE_CADASTRE.PMBC_PARCEL_FABRIC_POLY_SVW/wfs?service=WFS&REQUEST=GetFeature&VERSION=1.3.0&outputFormat=application/json&typeNames=pub:WHSE_CADASTRE.PMBC_PARCEL_FABRIC_POLY_SVW';

export const parcelLayerPopupConfig = {
  PARCEL_NAME: { label: 'Parcel Name:', display: (data: any) => data.PARCEL_NAME },
  PIN: { label: 'Parcel PIN:', display: (data: any) => data.PIN },
  PID: { label: 'Parcel PID:', display: (data: any) => data.PID },
  PLAN_NUMBER: { label: 'Plan Number:', display: (data: any) => data.PLAN_NUMBER },
  PARCEL_CLASS: { label: 'Parcel Class:', display: (data: any) => data.PARCEL_CLASS },
  MUNICIPALITY: { label: 'Municipality:', display: (data: any) => data.MUNICIPALITY },
  REGIONAL_DISTRICT: {
    label: 'Regional District:',
    display: (data: any) => data.REGIONAL_DISTRICT,
  },
  FEATURE_AREA_SQM: {
    label: 'Area:',
    display: (data: any) => (
      <>
        {data.FEATURE_AREA_SQM} m<sup>2</sup>
      </>
    ),
  },
};

export const municipalityLayerPopupConfig = {
  ADMIN_AREA_GROUP_NAME: {
    label: 'Administration Area:',
    display: (data: any) => `${data.ADMIN_AREA_GROUP_NAME} (${data.CHANGE_REQUESTED_ORG})`,
  },
};
