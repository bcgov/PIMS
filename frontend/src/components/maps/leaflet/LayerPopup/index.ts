export * from './LayerPopupContent';
export * from './constants';
export * from './hooks/useLayerQuery';
