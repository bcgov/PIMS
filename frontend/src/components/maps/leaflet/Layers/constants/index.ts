export const MUNICIPALITY_LAYER_URL =
  'https://openmaps.gov.bc.ca/geo/pub/WHSE_LEGAL_ADMIN_BOUNDARIES.ABMS_MUNICIPALITIES_SP/wfs?SERVICE=WFS&REQUEST=GetFeature&VERSION=1.3.0&outputFormat=application/json&typeNames=pub:WHSE_LEGAL_ADMIN_BOUNDARIES.ABMS_MUNICIPALITIES_SP';
export const PARCELS_LAYER_URL =
  'https://openmaps.gov.bc.ca/geo/pub/WHSE_CADASTRE.PMBC_PARCEL_FABRIC_POLY_SVW/wfs?service=WFS&REQUEST=GetFeature&VERSION=1.3.0&outputFormat=application/json&typeNames=pub:WHSE_CADASTRE.PMBC_PARCEL_FABRIC_POLY_SVW';

export const parcelLayerPopupConfig = {
  PIN: { label: 'Parcel PIN:', display: (data: any) => data.PIN },
  PID: { label: 'Parcel PID:', display: (data: any) => data.PID },
  REGIONAL_DISTRICT: {
    label: 'Regional District:',
    display: (data: any) => data.REGIONAL_DISTRICT,
  },
};

export const municipalityLayerPopupConfig = {
  ADMIN_AREA_NAME: { label: 'Area Name:', display: (data: any) => data.ADMIN_AREA_NAME },
  ADMIN_AREA_ABBREVIATION: {
    label: 'Area Abbrevation:',
    display: (data: any) => data.ADMIN_AREA_ABBREVIATION,
  },
  ADMIN_AREA_GROUP_NAME: {
    label: 'Area Group Name:',
    display: (data: any) => `${data.ADMIN_AREA_GROUP_NAME} (${data.CHANGE_REQUESTED_ORG})`,
  },
};
