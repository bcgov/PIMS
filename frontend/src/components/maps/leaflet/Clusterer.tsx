import React, { PropsWithChildren } from 'react';
import { LatLng, LeafletMouseEvent, Icon, DivIcon, Layer, Marker, GeoJSONEvent } from 'leaflet';
import { Map as LeafletMap, TileLayer, Popup, WMSTileLayer, GeoJSON } from 'react-leaflet';
import { BBox, GeoJsonProperties } from 'geojson';
import Supercluster from 'supercluster';
import useSupercluster, { ICluster } from 'hooks/useSupercluster';
import { createSingleMarker, createClusterMarker } from './mapUtils';
import { uniqueId } from 'lodash';

export type ClustererProps<P> = {
  points: Supercluster.PointFeature<P>[];
  bounds?: BBox;
  zoom: number;
  onMarkerClick?: Function;
  onClusterClick?: Function;
};

const Clusterer = <P extends GeoJsonProperties>(props: PropsWithChildren<ClustererProps<P>>) => {
  // state and refs
  const { points, bounds, zoom, onMarkerClick, onClusterClick } = props;

  // get clusters
  // clusters are an array of GeoJSON Feature objects, but some of them
  // represent a cluster of points, and some represent individual points.
  const { clusters, supercluster } = useSupercluster({
    points,
    bounds,
    zoom,
    options: { radius: 75, maxZoom: 18 },
  });

  const pointToLayer = (feature: ICluster, latlng: LatLng): Layer => {
    const { cluster: isCluster } = feature?.properties;
    // we have a cluster to render
    if (!!isCluster) {
      return createClusterMarker(feature, latlng);
    }
    // we have a single point (parcel or building) to render
    return createSingleMarker(feature, latlng);
  };

  const onClick = (e: LeafletMouseEvent) => {
    // the point may be either a cluster or a single map pin
    const cluster = (e?.layer as Marker)?.feature as ICluster;
    const isCluster = cluster?.properties?.cluster;

    // the user clicked on a cluster
    if (isCluster) {
      onClusterClick?.(cluster);
    } else {
      // the user clicked on a single map pin
      onMarkerClick?.(cluster);
    }
  };

  // TODO: improve typing
  return clusters && clusters?.length > 0 ? (
    <GeoJSON
      key={uniqueId('geojson_')}
      data={clusters as any}
      pointToLayer={pointToLayer}
      onclick={onClick}
    ></GeoJSON>
  ) : null;
};

export default Clusterer;
