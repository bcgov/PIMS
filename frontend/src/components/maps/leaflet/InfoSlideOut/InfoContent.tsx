import * as React from 'react';
import { ListGroup, Row } from 'react-bootstrap';
import { IBuilding, IParcel } from 'actions/parcelsActions';
import { Label } from 'components/common/Label';
import { ParcelPIDPIN } from './ParcelPIDPIN';
import ParcelAttributes from './ParcelAttributes';
import './InfoSlideOut.scss';
import BuildingAttributes from './BuildingAttributes';
import styled from 'styled-components';
import { ThreeColumnItem } from './ThreeColumnItem';
import variables from '_variables.module.scss';
import { PropertyTypes } from 'constants/propertyTypes';
import useCodeLookups from 'hooks/useLookupCodes';
import { useState } from 'react';
import { ProjectNumberLink } from './ProjectNumberLink';
import { BuildingSvg, LandSvg, SubdivisionSvg } from 'components/common/Icons';

/**
 * Compare two dates to evaluation which is earlier.
 * This should handle 'undefined' values by treating them as earlier.
 * @param a First date to compare.
 * @param b Second date to compare.
 * @returns A negative number indicating ealier, 0 as equal, positive number as later.
 */
export const compareDate = (a: Date | string | undefined, b: Date | string | undefined): number => {
  if (a === undefined && b === undefined) return 0;
  if (a === undefined && b !== undefined) return -1;
  if (a !== undefined && b === undefined) return 1;
  const aDate = typeof a === 'string' ? new Date(a) : (a as Date);
  const bDate = typeof b === 'string' ? new Date(b) : (b as Date);
  return aDate.valueOf() - bDate.valueOf();
};

interface IInfoContent {
  /** the selected property information */
  propertyInfo: IParcel | IBuilding | null;
  /** The property type [Parcel, Building] */
  propertyTypeId: PropertyTypes | null;
  /** whether the user has the correct agency/permissions to view all the details */
  canViewDetails: boolean;
}

export const OuterRow = styled(Row)`
  margin: 0px 0px 10px 0px;
`;

const ContactSres = styled(Row)`
  background-color: #deefff;
  padding: 5px 15px;
  margin: 0px 1px 5px 1px;

  em {
    color: ${variables.darkVariantColor};
  }

  a {
    color: ${variables.slideOutBlue};
  }
`;

const getHeading = (propertyTypeId: PropertyTypes | null) => {
  switch (propertyTypeId) {
    case PropertyTypes.SUBDIVISION:
      return (
        <Label className="header">
          <SubdivisionSvg className="svg" style={{ height: 25, width: 25, marginRight: 5 }} />
          Potential Subdivision
        </Label>
      );
    case PropertyTypes.BUILDING:
      return (
        <Label className="header">
          <BuildingSvg className="svg" style={{ height: 25, width: 25, marginRight: 5 }} />
          Building Identification
        </Label>
      );
    default:
      return (
        <Label className="header">
          <LandSvg className="svg" style={{ height: 25, width: 25, marginRight: 5 }} />
          Parcel Identification
        </Label>
      );
  }
};

/**
 * Component that displays the appropriate information about the selected property
 * in the property info slideout
 * @param {IInfoContent} propertyInfo the selected property
 * @param {IInfoContent} propertyTypeId the property type [Parcel, Building]
 * @param canViewDetails user can view all property details
 */
export const InfoContent: React.FC<IInfoContent> = ({
  propertyInfo,
  propertyTypeId,
  canViewDetails,
}) => {
  const isParcel =
    propertyTypeId !== null &&
    [PropertyTypes.PARCEL, PropertyTypes.SUBDIVISION].includes(propertyTypeId);

  const lookupCodes = useCodeLookups();
  const [privateProject, setPrivateProject] = useState<boolean>(false);

  return (
    <>
      <ListGroup>
        {getHeading(propertyTypeId)}
        {isParcel && <ParcelPIDPIN parcelInfo={propertyInfo as IParcel} />}
        <OuterRow>
          {canViewDetails && (
            <>
              {propertyInfo?.name && (
                <ThreeColumnItem leftSideLabel={'Name'} rightSideItem={propertyInfo?.name} />
              )}
              {propertyInfo?.subAgency ? (
                <>
                  <ThreeColumnItem
                    leftSideLabel={'Ministry'}
                    rightSideItem={lookupCodes.getAgencyFullName(propertyInfo?.agency)}
                  />
                  <ThreeColumnItem
                    leftSideLabel={'Owning agency'}
                    rightSideItem={lookupCodes.getAgencyFullName(propertyInfo.subAgency)}
                  />
                </>
              ) : (
                <ThreeColumnItem
                  leftSideLabel={'Owning ministry'}
                  rightSideItem={lookupCodes.getAgencyFullName(propertyInfo?.agency)}
                />
              )}
            </>
          )}
          <ThreeColumnItem
            leftSideLabel={'Classification'}
            rightSideItem={propertyInfo?.classification}
          />
          {!!propertyInfo?.projectNumbers?.length && (
            <ThreeColumnItem
              leftSideLabel={`Project Number${propertyInfo.projectNumbers.length > 1 ? 's' : ''}`}
              rightSideItem={propertyInfo.projectNumbers.map((projectNum: string) => (
                <ProjectNumberLink
                  key={projectNum}
                  setPrivateProject={setPrivateProject}
                  privateProject={privateProject}
                  agencyId={propertyInfo.agencyId}
                  projectNumber={projectNum}
                  breakLine
                />
              ))}
            />
          )}
        </OuterRow>
        {(!canViewDetails || privateProject) && (
          <ContactSres>
            <em>
              For more information on this property, contact{' '}
              <a href="mailto:RealPropertyDivision.Disposals@gov.bc.ca">
                Strategic Real Estate Services (SRES)
              </a>
            </em>
          </ContactSres>
        )}
      </ListGroup>
      <ListGroup>
        <Label className="header">Location data</Label>
        <OuterRow>
          <ThreeColumnItem
            leftSideLabel={'Civic address'}
            rightSideItem={propertyInfo?.address?.line1}
          />
          <ThreeColumnItem
            leftSideLabel={'Location'}
            rightSideItem={propertyInfo?.address?.administrativeArea}
          />
        </OuterRow>
        <OuterRow>
          <ThreeColumnItem leftSideLabel={'Latitude'} rightSideItem={propertyInfo?.latitude} />
          <ThreeColumnItem leftSideLabel={'Longitude'} rightSideItem={propertyInfo?.longitude} />
        </OuterRow>
      </ListGroup>
      {isParcel && (
        <ParcelAttributes parcelInfo={propertyInfo as IParcel} canViewDetails={canViewDetails} />
      )}
      {propertyTypeId === PropertyTypes.BUILDING && (
        <BuildingAttributes
          buildingInfo={propertyInfo as IBuilding}
          canViewDetails={canViewDetails}
        />
      )}
    </>
  );
};

export default InfoContent;
