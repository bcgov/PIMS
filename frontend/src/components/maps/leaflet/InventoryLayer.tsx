import React, { useMemo } from 'react';
import { IPropertyDetail } from 'actions/parcelsActions';
import { IGeoSearchParams } from 'constants/API';
import { BBox, Feature } from 'geojson';
import useDeepCompareEffect from 'hooks/useDeepCompareEffect';
import { LatLngBounds } from 'leaflet';
import { useLeaflet } from 'react-leaflet';
import { toast } from 'react-toastify';
import { PointFeature } from '../types';
import PointClusterer from './PointClusterer';
import { useApi } from 'hooks/useApi';
import _ from 'lodash';
import { useSelector } from 'react-redux';
import { RootState } from 'reducers/rootReducer';

export type InventoryLayerProps = {
  /** Latitude and Longitude boundary of the layer. */
  bounds: LatLngBounds;
  /** Zoom level of the map. */
  zoom: number;
  /** Minimum zoom level allowed. */
  minZoom?: number;
  /** Maximum zoom level allowed. */
  maxZoom?: number;
  /** Search filter to apply to properties. */
  filter?: IGeoSearchParams;
  /** What to do when the point feature is clicked. */
  onMarkerClick?: (point: PointFeature, position?: [number, number]) => void;

  selected?: IPropertyDetail | null;
};

/**
 * Get a new instance of a BBox from the specified 'bounds'.
 * @param bounds The latitude longitude boundary.
 */
const getBbox = (bounds: LatLngBounds): BBox => {
  return [
    bounds.getSouthWest().lng,
    bounds.getSouthWest().lat,
    bounds.getNorthEast().lng,
    bounds.getNorthEast().lat,
  ] as BBox;
};

/**
 * Get a new instance of a BBox from the specified 'bounds'.
 * @param bounds The latitude longitude boundary.
 */
const getApiBbox = (bounds: LatLngBounds): BBox => {
  return [
    bounds.getSouthWest().lng,
    bounds.getNorthEast().lng,
    bounds.getSouthWest().lat,
    bounds.getNorthEast().lat,
  ] as BBox;
};

/**
 * Displays the search results onto a layer with clustering.
 * This component makes a request to the PIMS API properties search WFS endpoint.
 */
export const InventoryLayer: React.FC<InventoryLayerProps> = ({
  bounds,
  zoom,
  minZoom,
  maxZoom,
  filter,
  onMarkerClick,
  selected,
}) => {
  const [features, setFeatures] = React.useState<Array<PointFeature>>([]);
  const { map } = useLeaflet();
  const { loadProperties } = useApi();
  const draftProperties: PointFeature[] = useSelector<RootState, PointFeature[]>(
    state => state.parcel.draftParcels,
  );

  if (!map) {
    throw new Error('<InventoryLayer /> must be used under a <Map> leaflet component');
  }

  const bbox = getBbox(bounds);

  minZoom = minZoom ?? 0;
  maxZoom = maxZoom ?? 18;

  const params = useMemo<IGeoSearchParams>(
    () => ({
      bbox: getApiBbox(bounds ?? map.getBounds()).toString(),
      address: filter?.address,
      administrativeArea: filter?.administrativeArea,
      pid: filter?.pid,
      projectNumber: filter?.projectNumber,
      agencies: filter?.agencies,
      classificationId: filter?.classificationId,
      minLandArea: filter?.minLandArea,
      maxLandArea: filter?.maxLandArea,
      inSurplusPropertyProgram: filter?.inSurplusPropertyProgram,
      inEnhancedReferralProcess: filter?.inEnhancedReferralProcess,
      floorCount: filter?.floorCount,
      predominateUseId: Number(filter?.predominateUseId),
      constructionTypeId: filter?.constructionTypeId,
      name: filter?.name,
      bareLandOnly: filter?.bareLandOnly,
      rentableArea: filter?.rentableArea,
    }),
    [filter, map, bounds],
  );

  const search = React.useCallback(
    _.debounce(
      (filter: IGeoSearchParams) => {
        loadProperties(filter)
          .then(async (data: Feature[]) => {
            const points = data
              .filter(feature => {
                return !(
                  feature.properties!.propertyTypeId === selected?.propertyTypeId &&
                  feature.properties!.id === selected?.parcelDetail?.id
                );
              })
              .map(f => {
                return {
                  ...f,
                } as PointFeature;
              });
            setFeatures(points);
          })
          .catch(error => {
            toast.error((error as Error).message, { autoClose: 7000 });
            console.error(error);
          });
      },
      500,
      { leading: true },
    ),
    [],
  );

  // Fetch the geoJSON collection of properties.
  useDeepCompareEffect(() => {
    search(params);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [params, bbox, selected, map]);

  return (
    <PointClusterer
      points={features}
      draftPoints={draftProperties}
      zoom={zoom}
      bounds={bbox}
      onMarkerClick={onMarkerClick}
      zoomToBoundsOnClick={true}
      spiderfyOnMaxZoom={true}
      selected={selected}
    />
  );
};
