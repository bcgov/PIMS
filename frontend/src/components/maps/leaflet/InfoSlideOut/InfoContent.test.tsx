import * as React from 'react';
import 'leaflet';
import 'leaflet/dist/leaflet.css';
import InfoContent from './InfoContent';
import { IParcel, IBuilding } from 'actions/parcelsActions';
import { PropertyTypes } from 'constants/propertyTypes';
import { render } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as reducerTypes from 'constants/reducerTypes';
import { Provider } from 'react-redux';
import * as API from 'constants/API';

const mockParcelNoSub = {
  id: 1,
  pid: '000-000-000',
  zoning: '',
  zoningPotential: '',
  classificationId: 1,
  encumbranceReason: '',
  agencyId: '',
  isSensitive: false,
  latitude: 48,
  longitude: 123,
  classification: 'Core Operational',
  name: 'test name',
  description: 'test',
  assessedLand: 10000,
  assessedBuilding: 10000,
  evaluations: [
    {
      date: new Date(),
      key: 'Assessed',
      value: 10000,
    },
  ],
  fiscals: [
    {
      fiscalYear: 2020,
      key: 'NetBook',
      value: 10000,
    },
  ],
  address: {
    id: 1,
    line1: '1234 mock Street',
    administrativeArea: 'Victoria',
    province: 'BC',
    postal: 'V1V1V1',
    provinceId: '1',
  },
  landArea: 123,
  landLegalDescription: 'test description',
  buildings: [],
  parcels: [],
  agency: 'AEST',
} as IParcel;

export const mockBuilding = {
  name: 'test name',
  id: 100,
  parcelId: '',
  pid: '',
  address: {
    id: 1,
    line1: '1234 mock Street',
    administrativeArea: 'Victoria',
    province: 'BC',
    postal: 'V1V1V1',
    provinceId: '1',
  },
  latitude: 48,
  longitude: 123,
  buildingConstructionTypeId: 0,
  buildingConstructionType: 'Concrete',
  classificationId: 1,
  classification: 'Core Operational',
  assessedLand: 10000,
  assessedBuilding: 10000,
  evaluations: [
    {
      date: new Date(),
      key: 'Assessed',
      value: 10000,
    },
  ],
  fiscals: [
    {
      fiscalYear: 2020,
      key: 'NetBook',
      value: 10000,
    },
  ],
  rentableArea: 100,
  totalArea: 200,
  agency: 'AEST',
  agencyId: 0,
  agencyCode: 'KPU',
  subAgency: 'KPU',
  transferLeaseOnSale: false,
  isSensitive: false,
  buildingPredominateUse: 'University/College',
  buildingPredominateUseId: 0,
  buildingOccupantTypeId: 0,
  encumbranceReason: '',
  occupantName: 'test',
  parcels: [mockParcelNoSub],
  buildingTenancy: '100%',
} as IBuilding;

export const mockParcel = {
  id: 1,
  pid: '000-000-000',
  pin: '',
  projectNumber: '',
  zoning: '',
  zoningPotential: '',
  classificationId: 1,
  encumbranceReason: '',
  agencyId: '',
  isSensitive: false,
  latitude: 48,
  longitude: 123,
  classification: 'Core Operational',
  name: 'test name',
  description: 'test',
  assessedLand: 10000,
  assessedBuilding: 10000,
  evaluations: [
    {
      date: new Date(),
      key: 'Assessed',
      value: 10000,
    },
  ],
  fiscals: [
    {
      fiscalYear: 2020,
      key: 'NetBook',
      value: 10000,
    },
  ],
  address: {
    id: 1,
    line1: '1234 mock Street',
    line2: 'N/A',
    administrativeArea: 'Victoria',
    province: 'BC',
    postal: 'V1V1V1',
    provinceId: '1',
  },
  landArea: 123,
  landLegalDescription: 'test description',
  buildings: [mockBuilding],
  parcels: [],
  agency: 'AEST',
  subAgency: 'KPU',
} as IParcel;

const lCodes = {
  lookupCodes: [
    {
      code: 'AEST',
      id: 1,
      isDisabled: false,
      name: 'Ministry of Advanced Education',
      type: API.AGENCY_CODE_SET_NAME,
    },
    {
      code: 'KPU',
      id: 181,
      isDisabled: false,
      name: 'Kwantlen Polytechnic University',
      type: API.AGENCY_CODE_SET_NAME,
    },
  ],
};

const mockStore = configureMockStore([thunk]);
const store = mockStore({
  [reducerTypes.LOOKUP_CODE]: lCodes,
});

const ContentComponent = (
  propertyInfo: IParcel | IBuilding | null,
  propertyTypeId: PropertyTypes | null,
  canViewDetails: boolean,
) => {
  return (
    <Provider store={store}>
      <InfoContent
        propertyInfo={propertyInfo}
        propertyTypeId={propertyTypeId}
        canViewDetails={canViewDetails}
      />
    </Provider>
  );
};

describe('InfoContent View', () => {
  it('InfoContent renders correctly', () => {
    const { container } = render(ContentComponent(mockParcel, PropertyTypes.PARCEL, true));
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Shows all parcel information when can view', () => {
    const { getByText } = render(ContentComponent(mockParcel, PropertyTypes.PARCEL, true));
    expect(getByText('Parcel Identification')).toBeVisible();
    //Identification information
    expect(getByText('000-000-000')).toBeVisible();
    expect(getByText('test name')).toBeVisible();
    expect(getByText('Ministry of Advanced Education')).toBeVisible();
    expect(getByText('Kwantlen Polytechnic University')).toBeVisible();
    expect(getByText('Core Operational')).toBeVisible();
    //Location data
    expect(getByText('1234 mock Street')).toBeVisible();
    expect(getByText('Victoria')).toBeVisible();
    expect(getByText('48')).toBeVisible();
    //Legal Description
    expect(getByText('test description')).toBeVisible();
  });

  it('Lot size formats correctly', () => {
    const { getByText } = render(ContentComponent(mockParcel, PropertyTypes.PARCEL, true));
    expect(getByText('123 hectares')).toBeVisible();
  });

  it('Assessed value formats correctly', () => {
    const { getByText } = render(ContentComponent(mockParcel, PropertyTypes.PARCEL, true));
    expect(getByText('$10,000')).toBeVisible();
  });

  it('Shows limited parcel information when cannot view', () => {
    const { getByText, queryByText } = render(
      ContentComponent(mockParcel, PropertyTypes.PARCEL, false),
    );
    expect(queryByText('test name')).toBeNull();
    expect(queryByText('Ministry of Advanced Education')).toBeNull();
    expect(queryByText('Kwantlen Polytechnic University')).toBeNull();
    //contact SRES block is shown
    expect(getByText('For more information', { exact: false })).toBeVisible();
  });

  it('Correct label if no sub agency', () => {
    const { getByText } = render(ContentComponent(mockParcelNoSub, PropertyTypes.PARCEL, true));
    expect(getByText('Owning ministry')).toBeVisible();
  });

  it('Shows all building information when can view', () => {
    const { getByText } = render(ContentComponent(mockBuilding, PropertyTypes.BUILDING, true));
    expect(getByText('Building Identification')).toBeVisible();
    //Identification information
    expect(getByText('test name')).toBeVisible();
    expect(getByText('Ministry of Advanced Education')).toBeVisible();
    expect(getByText('Kwantlen Polytechnic University')).toBeVisible();
    expect(getByText('Core Operational')).toBeVisible();
    //Location data
    expect(getByText('1234 mock Street')).toBeVisible();
    expect(getByText('Victoria')).toBeVisible();
    expect(getByText('48')).toBeVisible();
    //Building Attributes
    expect(getByText('University/College')).toBeVisible();
    expect(getByText('100%')).toBeVisible();
  });

  it('Building area formated correctly', () => {
    const { getByText } = render(ContentComponent(mockBuilding, PropertyTypes.BUILDING, true));
    expect(getByText('100 sq. metres')).toBeVisible();
  });

  it('Shows limited building information when cannot view', () => {
    const { getByText, queryByText } = render(
      ContentComponent(mockBuilding, PropertyTypes.BUILDING, false),
    );
    expect(queryByText('test name')).toBeNull();
    expect(queryByText('Ministry of Advanced Education')).toBeNull();
    expect(queryByText('Kwantlen Polytechnic University')).toBeNull();
    //contact SRES block is shown
    expect(getByText('For more information', { exact: false })).toBeVisible();
  });
});
