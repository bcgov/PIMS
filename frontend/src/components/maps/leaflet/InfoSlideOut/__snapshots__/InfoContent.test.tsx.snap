// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoContent View InfoContent renders correctly 1`] = `
.c1 {
  margin: 0px;
  width: 300px;
}

.c2 {
  width: 110px;
  max-width: 110px;
  padding-right: 10px;
  padding-left: 0px;
}

.c3 {
  max-width: 1px;
  padding: 0px;
  background-color: rgba(96,96,96,0.2);
}

.c4 {
  padding-left: 10px;
  padding-right: 0px;
}

.c0 {
  margin: 0px 0px 10px 0px;
}

<div
  class="list-group"
>
  <p
    class="label header"
  >
    <svg
      class="svg"
      style="height: 25px; width: 25px; margin-right: 5px;"
    >
      icon-lot.svg
    </svg>
    Parcel Identification
  </p>
  <div
    class="c0 row"
  >
    <div
      class="c1 row"
    >
      <div
        class="c2 col"
      >
        <div
          class="left-side list-group-item"
        >
          <p
            class="label"
          >
            PID
          </p>
        </div>
      </div>
      <div
        class="c3 col"
      />
      <div
        class="c4 col"
      >
        <div
          class="list-group-item"
        >
          000-000-000
        </div>
      </div>
    </div>
    
  </div>
  <div
    class="c0 row"
  >
    <div
      class="c1 row"
    >
      <div
        class="c2 col"
      >
        <div
          class="left-side list-group-item"
        >
          <p
            class="label"
          >
            Name
          </p>
        </div>
      </div>
      <div
        class="c3 col"
      />
      <div
        class="c4 col"
      >
        <div
          class="list-group-item"
        >
          test name
        </div>
      </div>
    </div>
    <div
      class="c1 row"
    >
      <div
        class="c2 col"
      >
        <div
          class="left-side list-group-item"
        >
          <p
            class="label"
          >
            Ministry
          </p>
        </div>
      </div>
      <div
        class="c3 col"
      />
      <div
        class="c4 col"
      >
        <div
          class="list-group-item"
        >
          Ministry of Advanced Education
        </div>
      </div>
    </div>
    <div
      class="c1 row"
    >
      <div
        class="c2 col"
      >
        <div
          class="left-side list-group-item"
        >
          <p
            class="label"
          >
            Owning agency
          </p>
        </div>
      </div>
      <div
        class="c3 col"
      />
      <div
        class="c4 col"
      >
        <div
          class="list-group-item"
        >
          Kwantlen Polytechnic University
        </div>
      </div>
    </div>
    <div
      class="c1 row"
    >
      <div
        class="c2 col"
      >
        <div
          class="left-side list-group-item"
        >
          <p
            class="label"
          >
            Classification
          </p>
        </div>
      </div>
      <div
        class="c3 col"
      />
      <div
        class="c4 col"
      >
        <div
          class="list-group-item"
        >
          Core Operational
        </div>
      </div>
    </div>
  </div>
</div>
`;
